### DISCUSSION TITLE
How to override `jpeg_factory` from `JpegImagePlugin`

### DISCUSSION THREAD

**Discussion by cmahnke** [python-pillow/Pillow#8269] [General]
Currently it's not possible to load a 'raw' as in unmodified JPEG file if Pillow suspects a file is a MPO, this is due to the fact that `jpeg_factory` in `JpegImagePlugin` contains a switch if it might be a MPO / UltraHDR file, is there a way to disable these checks 
and work with those files?

Jus reregistring a open handler with my own function like this won't work:
```python
Image.register_open(JpegImageFile.format, jpeg_factory, JpegImagePlugin._accept)
```

Manipulating the registry (except adding things) isn't exposed via API either.

<s>BTW: Detecting images produced by `libultrahdr` doesn't seem to work for images build with recent versions of the lib build from source. They are still identified as MPO.</s>
Please, ignore the last comment

**Comment by cmahnke** [comment]
Ok, got the first issue solved, it was an issue with the loading order...

**Comment by Yay295** [comment]
An `MpoImageFile` *is* a `JpegImageFile`. What is the actual problem you are having?

**Reply by cmahnke** [reply]
I'm trying to build a Handler for UHDR images...

**Reply by Yay295** [reply]
I'm not sure how that changes anything, but have you tried monkey patching `jpeg_factory`?

`JpegImagePlugin.jpeg_factory = your_function`

**Reply by radarhere** [reply]
What version of Pillow are you using? I would have thought this would be resolved by #8056

**Reply by cmahnke** [reply]
@Yay295: Yes, this works.
@radarhere: It is, but in a way that makes it hard to integrate something if there is a UHDR Image...

Currently I'm struggling with setting XMP data to be written - Pillow has some asymetric APIs.
From what I understand, it should be possible to pass binary data as `info["extra"]` ([see](https://github.com/python-pillow/Pillow/blob/4df4df2a18ded44811be786703a43c3460cb3bc9/src/PIL/JpegImagePlugin.py#L754)) to the image and `JpegImagePlugin._save` will prepend it to the ICC profile.
Has anyone tried this?

**Reply by radarhere** [reply]
An example of `"extra"` being used is in MpoImagePlugin - https://github.com/python-pillow/Pillow/blob/4df4df2a18ded44811be786703a43c3460cb3bc9/src/PIL/MpoImagePlugin.py#L51-L62

**Reply by radarhere** [reply]
If you're working on a pull request to add support for Ultra HDR images, then feel free to revert parts of #8056

**Reply by cmahnke** [reply]
Thanks, I'll have a look.

But it seems to me that the provided example will overwrite any passed `extra`, so passing just some binary encoded XMP would get lost there - I need to create my own `save_all` function.

I don't think this will end up in a pull request, since it's currently quite hacky:
* I did the obvious, which is `class UhdrImageFile(MpoImagePlugin.MpoImageFile)`. But this required me to duplicate `seek()` and `tell()` since `self.__frame` is "private" (two underscores in front) - A clean solution might require changes to `MpoImageFile` as well. But currently I don't know if my changes are really needed.
* My `UhdrImageFile` is currently basically just a container, every real operation (like `crop`, `resize` and also `save`) is done using functions not part of the class
* For XMP manipulations I did't do it the pythonic way, which would be a `dict` based interface, I used `str` and from there: 
* For XMP manipulations I depend on `ElementTree`'s XPATH capabilities - Pillow uses another XML lib

But maybe it's worth to implement a XMP write solution, since this is asked for more often...

**Reply by cmahnke** [reply]
Ok, XMP itself is a whole other can of worms: One key seems to be #5076, I would like to have a proper XML interface parallel to the existing `dict` based `getxmp()` method, just because it's easier to work on data that isn't an abstraction...
I would also ignore the possibility to have the XMP data split over multiple fields in the first try... 

The first part, adding XMP to JPEG was far easier then expected:
```python
from PIL._binary import o16be

XMP_MARKER = b"http://ns.adobe.com/xap/1.0/"
XMP_MAX = 65504

def xmp_extra(xml: str):
    if len(xml) > XMP_MAX:
        raise
    length = len(XMP_MARKER) + 2 + len(xml) + 1
    return b"\xFF\xE1" + o16be(length) + XMP_MARKER + b"\0" + xml.encode()

image.encoderinfo['extra'] = xmp_extra(xml)
```

This gives just a minor warning with `exiftool`:
```
exiftool -validate -warning -error -a ../img/xmp.jpeg 
Validate                        : 1 Warning (minor)
Warning                         : [minor] XMP is missing xpacket wrapper
```
But this also happens with a file generated by the `libultrahdr` app.

Of course without handling XML larger then 64KB
So easy in fact, that I expect some other pitfalls, otherwise it would be possible with Pillow for years 😉

Next step is to get this into a fork of the `MpoImagePlugin._save_all` method

**Reply by cmahnke** [reply]
Well, I'm currently stuck. I have problems figuring out how the MPO writer actually works - I suspect some constants are jus magic numbers? Or better: Field length don't seem to be calculated based on the given input. Trying to pass just XMP as prefix using encoderinfo['extra'] results in a mixup of the MP Part and the ICC Profile...

**Reply by cmahnke** [reply]
@radarhere : Thanks for #8286. It actually helped me to get a bit further, I hab been blind and didn't noticed that the `Image.save()`method resets `encoderinfo` from the image. Anyways, I'm able to add a XMP field to secondary image now, but UltraHDR has another requirement I need to provide:
```
apply gainmap method expects base image color format to be one of {UHDR_IMG_FMT_24bppYCbCr444, UHDR_IMG_FMT_16bppYCbCr422, UHDR_IMG_FMT_12bppYCbCr420}. Received 2
```

I suspect to be able to figure it out, just running `convert('YCbCr')` on the primary image (the one that the `append_images` array is added to), doesn't seem to do the trick for now.
